internal/cmdrpkgfindupdates/command.go

```diff
@@ -0,0 +1,221 @@
+		//Short:   rpkgdocs.FindUpdatesShort,
+		//Long:    rpkgdocs.FindUpdatesShort + "\n" + rpkgdocs.FindUpdatesLong,
+		//Example: rpkgdocs.FindUpdatesExamples,
+		PreRunE: r.preRunE,
+		RunE:    r.runE,
+		Hidden:  porch.HidePorchCommands,
+	}
+	r.Command = c
+	return r
+}
+
+type runner struct {
+	ctx     context.Context
+	cfg     *genericclioptions.ConfigFlags
+	client  client.Client
+	Command *cobra.Command
+
+	// Flags
+}
+
+func (r *runner) preRunE(_ *cobra.Command, args []string) error {
+	const op errors.Op = command + ".preRunE"
+
+	if len(args) < 1 {
+		return errors.E(op, "PACKAGE_REVISION is a required positional argument")
+	}
+	cf, err := r.cfg.ToRESTConfig()
+	if err != nil {
+		return errors.E(op, err)
+	}
+	scheme, err := rpkgutil.CreateScheme()
+	if err != nil {
+		return errors.E(op, err)
+	}
+	c, err := client.New(cf, client.Options{Scheme: scheme})
+	if err != nil {
+		return errors.E(op, err)
+	}
+	r.client = c
+	return nil
+}
+
+func (r *runner) runE(cmd *cobra.Command, args []string) error {
+	const op errors.Op = command + ".runE"
+	var messages []string
+
+	for _, p := range args {
+		packageRevision, err := r.getPackageRevision(p)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		upstreamLock, err := r.getUpstreamLock(packageRevision)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		availableUpdates, err := r.availableUpdates(upstreamLock)
+		if err != nil {
+			return errors.E(op, err)
+		}
+
+		fmt.Fprintf(cmd.OutOrStdout(), "%s %s\n", p, availableUpdates)
+	}
+
+	if len(messages) > 0 {
+		return errors.E(op, fmt.Errorf("errors:\n  %s", strings.Join(messages, "\n  ")))
+	}
+
+	return nil
+}
+
+func (r *runner) getPackageRevision(p string) (porchapi.PackageRevision, error) {
+	packageRevision := porchapi.PackageRevision{}
+	newScheme := runtime.NewScheme()
+	if err := porchapi.SchemeBuilder.AddToScheme(newScheme); err != nil {
+		return packageRevision, err
+	}
+	restClient, err := porch.CreateRESTClient(r.cfg)
+	if err != nil {
+		return packageRevision, err
+	}
+	err = restClient.
+		Get().
+		Namespace(*r.cfg.Namespace).
+		Resource("packagerevisions").
+		Name(p).
+		Do(r.ctx).
+		Into(&packageRevision)
+
+	return packageRevision, err
+}
+
+func (r *runner) getUpstreamLock(packageRevision porchapi.PackageRevision) (*kptfile.UpstreamLock, error) {
+	var resources porchapi.PackageRevisionResources
+	if err := r.client.Get(r.ctx, client.ObjectKey{
+		Namespace: *r.cfg.Namespace,
+		Name:      packageRevision.Name,
+	}, &resources); err != nil {
+		return nil, err
+	}
+
+	// get the upstream package URL from the Kptfile
+	var kf *kptfile.KptFile
+	for k, v := range resources.Spec.Resources {
+		if k == kptfile.KptFileName {
+			var err error
+			kf, err = pkg.DecodeKptfile(strings.NewReader(v))
+			if err != nil {
+				return nil, err
+			}
+		}
+	}
+
+	return kf.UpstreamLock, nil
+}
+
+func (r *runner) availableUpdates(upstreamLock *kptfile.UpstreamLock) ([]string, error) {
```

Comment:
I think it might be a better solution here to check for newer revisions of packages through the porch api rather than going to the git repo directly. I think we can assume that all later package revisions of the same package as the current upstream are eligible.

Reply:
+1 to what @mortent said

> but then thought the upstream repo might not be registered with porch

I think narrowing down the scope to only limit registered repo is a fine starting point. unregistered repo introduces lot of edge cases and some scenario that are difficult to reason with.

Reply:
Got it. I will update accordingly.